presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java

```diff
@@ -223,18 +240,137 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
         }
         else {
             return new PlanBuilder(translations,
-                    // TODO handle parameter list
                     new ApplyNode(idAllocator.getNextId(),
                             root,
                             subqueryNode,
-                            ImmutableList.of()),
+                            ImmutableList.copyOf(correlation.values())),
                     subPlan.getSampleWeight());
         }
     }
 
+    private Map<Expression, Symbol> extractCorrelation(PlanBuilder subPlan, PlanNode subquery)
+    {
+        Set<Expression> missingReferences = extractMissingExpressions(subquery);
+        ImmutableMap.Builder<Expression, Symbol> correlation = ImmutableMap.builder();
+        if (!missingReferences.isEmpty()) {
+            for (Expression missingReference : missingReferences) {
+                Expression rewritten = subPlan.rewrite(missingReference);
+                if (rewritten != missingReference) {
+                    correlation.put(missingReference, Symbol.from(rewritten));
+                }
+            }
+        }
+        return correlation.build();
+    }
+
     private RelationPlan createRelationPlan(Query subquery)
     {
         return new RelationPlanner(analysis, symbolAllocator, idAllocator, metadata, session)
                 .process(subquery, null);
     }
+
+    private Set<Expression> extractMissingExpressions(PlanNode planNode)
```

Comment:
I will rename it. It finds missing column reference expressions i.e. column reference which cannot be satisfied within given PlanNode.

At this point all column reference should be rewritten to symbol references. If it finds column reference that means this is a reference to column from outer scope